add char support

FreeMasen opened the pull request:
this should address #198 
I have an example project I used to test this is working. I wasn't sure if it should be included.

alexcrichton commented:
Thanks for the PR! 

I think the values related to `char` on `JsValue` can be removed though in the sense that `JsValue` is attempting to be as close to a native JS value as it can. In that sense JS doesn't actually have a native character type, so it should be ok to remove.

Other than that mind adding some tests to exercise this functionality? Looks great other than those nits!

FreeMasen commented:
I will add the example project I used to test this working and remove the `JsValue` entries. I also appears that I need to move the changes into the `if_std!` block?

alexcrichton commented:
I think the import just needs to be from `core::char` instead of `std::char` and should otherwise be good to go

alexcrichton commented on src/lib.rs:
```diff
@@ -305,6 +305,8 @@ externs! {
     fn __wbindgen_boolean_new(val: u32) -> u32;
     fn __wbindgen_boolean_get(idx: u32) -> u32;
+    fn __wbindgen_char_new(val: u32) -> u32;
+    fn __wbindgen_char_get(idx: u32) -> u32;
```
I think this can be removed since the `JsValue` methods were removed

alexcrichton commented on crates/cli-support/src/js/mod.rs:
```diff
@@ -216,6 +216,25 @@ impl<'a> Context<'a> {
+                    return 0;
+                }}
+            "))
+        })?;
```
I think this can be removed since the `JsValue` methods were removed

alexcrichton commented on tests/all/char.rs:
```diff
@@ -0,0 +1,42 @@
+            pub fn wide_char() -> char { '💩' }
+
+            #[wasm_bindgen]
+            pub fn parrot(c: char) -> char { c }
```
Could this be expanded to also assert the value of the character coming in to Rust?

alexcrichton commented on examples/char/.gitignore:
```diff
@@ -0,0 +1,4 @@
+package-lock.json
+char.js
+char_bg.js
+char_bg.wasm
```
Could this ignore `dist` as well? (I think a 1.2MB js file was checked in below)

FreeMasen commented:
I added a `README.md` for the example and updated the examples `README.md`. I trimmed the char project a little bit as well. I started to try and create a version in `WebAssembly.studio` but that would only be possible once these changes are merged. 

I also removed the large `js` file from the repo.

Once this is all ready did you want me to squash the interim commits into 'add char support'?

alexcrichton commented:
Looks great! And no need, nowadays with GitHub you can squash at merge time!

kimtuck commented:
@FreeMasen @alexcrichton Thanks for the fix.  I'm working on some samples using bindgen for a presentation; I could upload the examples someplace (via pull request) if there is a place here that makes sense

alexcrichton commented:
@kimtuck certainly! Either in this repo under `examples` or gists on https://webassembly.studio/ would be great!
